src/intro.tex

```diff
@@ -165,6 +165,15 @@ \section{RISC-V Software Execution Environments and Harts}
 Some EEIs support the creation and destruction of additional harts,
 for example, via environment calls to fork new harts.
 
+The execution environment is responsible for ensuring eventual
+forward progress of each of its harts.  Forward progress includes:
+\begin{itemize}
+\item Completing the execution of an instruction.
+\item Other events defined by extensions, such as traps.
+\item Mechanisms that explicitly wait for an event, such as the WFI
+  instruction waiting for interrupts, or various environment calls.
```

Comment:
I like the first two.  I parse the third as "Mechanisms that explicitly wait for ... various environment calls."  What environment calls do you envision here?  Wouldn't ECALL instructions fall under the first bullet?

Are mechanisms that wait for an event (which may never come) really considered to be making forward progress (as stated here) or are we waiving the forward progress guarantee in this case?  My intuition is the latter so it would be more like:
Unless a mechanism that explicitly waits for an event, such as the WFI instruction waiting for interrupts, is being employed, the execution environment is responsible for ensuring eventual forward progress of each of its harts.  Forward progress includes:
- Completing the execution of an instruction.
- Other events defined by extensions, such as traps.

Reply:
Good points.  I did ponder a bit about whether to make "waiting" part of forward progress or an exception to it.  Your alternative looks good.